experimental/torch_xla2/examples/train_llama_torchtitan/train_llama.py

```diff
@@ -0,0 +1,283 @@
+  # only need for tpu v4
+  # os.environ.setdefault('TPU_MEGACORE', 'megacore_dense')
+  tpu_args = "--xla_tpu_enable_async_collective_fusion_fuse_all_gather=true --xla_tpu_megacore_fusion_allow_ags=false --xla_enable_async_collective_permute=true --xla_tpu_enable_ag_backward_pipelining=true --xla_tpu_enable_data_parallel_all_reduce_opt=true --xla_tpu_data_parallel_opt_different_sized_ops=true --xla_tpu_enable_async_collective_fusion=true --xla_tpu_enable_async_collective_fusion_multiple_steps=true --xla_tpu_overlap_compute_collective_tc=true --xla_enable_async_all_gather=true"
+  os.environ.setdefault('LIBTPU_INIT_ARGS', tpu_args)
+
+_setup_default_env()
+
+
+
+SEQLEN = 8192
+BATCH = 8
+global_axis: Tuple[str, str] = ('fsdp', )
+num_global_devices = jax.device_count()
+num_local_devices = jax.local_device_count()
+num_partitions = (num_global_devices, )
+
+
+def sharded_device_put(tensor, sharding):
+    if isinstance(tensor, tuple):
+        return tuple(sharded_device_put(t, sharding) for t in tensor)
+
+    if num_global_devices == num_local_devices:
+        return jax.device_put(tensor, sharding)
+
+    shape = tensor.shape
+    x_split = [jax.device_put(tensor[i], device) for device, i in sharding.addressable_devices_indices_map(shape).items()]
+    return jax.make_array_from_single_device_arrays(shape, sharding, x_split)
+
+
+class FSDPv2(torch.nn.Module):
+
+    def __init__(self, mod):
+        super().__init__()
+        self.mod = mod
+        self.mesh = jax.sharding.Mesh(
+            mesh_utils.create_device_mesh(num_partitions),
+            axis_names=global_axis,
+        )
+        self.sharding = jax.sharding.NamedSharding(self.mesh, P(*global_axis))
+
+    def forward(self, *args):
+        args = list(args)
+        args[0] = self.shard(args[0])
+        res = self.mod(*args)
+        return self.shard(res)
+    
+    def shard(self, x):
+        return torch_xla2.interop.call_jax(
+            jax.lax.with_sharding_constraint,
+            x,
+            self.sharding,
+        )
+
+def print_shapes(pyt):
+    for p in pytree.tree_flatten(pyt)[0]:
+        if hasattr(p, 'shape'):
+            print(p.shape, p.dtype)
+
+
+class Module(torch.nn.Module):
+
+    def __init__(self, inner):
+        super().__init__()
+        self.inner = FSDPv2(inner)
+
+    def training_step(self, data, batch_id):
+        x, y = data
+        logits = self.inner(x)
+        num_tokens = logits.shape[-1]
+        logits = logits.reshape(-1, num_tokens)
+        y = y.reshape(-1)
+        return torch.nn.functional.cross_entropy(
+            logits, y)
+    
+
+class Trainer:
+
+    def __init__(self):
+        self.mesh = jax.sharding.Mesh(
+            mesh_utils.create_device_mesh(num_partitions),
+            axis_names=global_axis,
+        )
+        self.x_sharding = jax.sharding.NamedSharding(self.mesh, P(global_axis))
+        self.y_sharding = jax.sharding.NamedSharding(self.mesh, P(*global_axis))
```

Comment:
I guess I wasn't sure how's `y_sharding` used. Maybe it's only relevant in a 2-D sharding world? Consider just removing this since this examples looks like specialized fro FSDP.

Reply:
done.